lib/spack/spack/spec.py

```diff
@@ -2454,7 +2454,7 @@ def _constrain_dependencies(self, other):
         """Apply constraints of other spec's dependencies to this spec."""
         other = self._autospec(other)
 
-        if not self._dependencies or not other._dependencies:
+        if not other._dependencies:
```

Comment:
Honestly I made the change months ago as I started working on this... I don't recall exactly how it failed. I had the same response you did when I first found it though.

Do you need me to try to go back and figure out what problem it caused?

Reply:
Yes please. If a test was added, then I assume reverting this line will be sufficient to uncover the problem. Otherwise a test should probably be added for it (and I'd be curious to know about the case that tripped over this).

Reply:
Adding a line in an existing constraint test for this.

The case that failed is `Spec('foo').constrain(Spec('^bar'))`, so it's a pretty basic thing that should've worked.